### ISSUE TITLE
QUIC: Unreliable Datagram Extension (RFC 9221) implementation

### ISSUE THREAD

**Issue by cyanreg** [openssl/openssl#24003]
Hi all,
I've started work on implementing QUIC support in [AVTransport](https://github.com/cyanreg/avtransport/), using the newly implemented/staged client/server API.

It is a multimedia transmission protocol (to become a part of FFmpeg), which may use FEC to fill in missing packets. Latency requirements make it impractical to retransmit packets when streaming real-time video, hence RFC 9221 introduces unreliable datagram streams. Currently, the OpenSSL QUIC API lacks handling for them, so I'm opening this issue as a tracker and a list of requirements.

Externally, to API users, handling shouldn't be difficult: introduce a new flag, like `SSL_STREAM_FLAG_UNRELIABLE`, to be given to `SSL_new_stream()`. Extra parameters, like adjusting retransmission behaviour, and ACK delaying would also significantly help.

Internally, and optimally, the entirety of the Frame-in-Flight Management (FIFM) should be able to be fully disabled (optionally, given user settings) for such streams. This would mean no reordering, no state keeping, and no buffering.
A goal is to enable 4k streaming at 20Gbps over QUIC, so reducing every bit of overhead helps.

Currently, all QUIC implementations which implement RFC 9221 do *some level* of management, even for unreliable streams, which leads to dropped, or delayed packets. OpenSSL is more flexible in general when it comes to letting users manage connections, so supporting these would go a long way in making this implementation the most optimal for real-time streaming.

As an aside, would it be possible to extend the BIO API to allow gather for datagrams when writing? Currently, the reference library's entire data path, from a user giving codec packet data to writing it out is entirely zero-copy. Using BIO requires that the datagram contain a single contiguous buffer, which would mean we would need to copy the user data onto a buffer, along with the header the protocol inserts, just so we can pass it off to OpenSSL (which then encodes the data to the output).
I am using datagram memory BIO as the reference library implements multiple I/O backends, from standard sendmsg/recvmsg, to using zero-copy liburing (WIP).

**Comment by mattcaswell** [comment]
@hlandau

**Comment by hlandau** [comment]
> Hi all, I've started work on implementing QUIC support in [AVTransport](https://github.com/cyanreg/avtransport/), using the newly implemented/staged client/server API.
> 
> It is a multimedia transmission protocol (to become a part of FFmpeg), which may use FEC to fill in missing packets. Latency requirements make it impractical to retransmit packets when streaming real-time video, hence RFC 9221 introduces unreliable datagram streams. Currently, the OpenSSL QUIC API lacks handling for them, so I'm opening this issue as a tracker and a list of requirements.

Indeed, there will be an API designed in future to support RFC 9221. There's a fair amount of possible design space there, so any input on API requirements is valuable.

> Externally, to API users, handling shouldn't be difficult: introduce a new flag, like `SSL_STREAM_FLAG_UNRELIABLE`, to be given to `SSL_new_stream()`. Extra parameters, like adjusting retransmission behaviour, and ACK delaying would also significantly help.

This probably isn't the right API approach. Would the expectation here be that on the created "stream", each SSL_write() call sends one datagram (ala UDP send())? In that case you are forced to create a new "singleton" stream object just to send datagrams. It might make sense if you want to have different classes of "datagram service" but otherwise I'm intending to go with a separate API call directly on a QUIC connection SSL object (`SSL_write_datagram` or similar). 
 
> Internally, and optimally, the entirety of the Frame-in-Flight Management (FIFM) should be able to be fully disabled (optionally, given user settings) for such streams. This would mean no reordering, no state keeping, and no buffering. A goal is to enable 4k streaming at 20Gbps over QUIC, so reducing every bit of overhead helps.

Obviously, since the datagram extension is about enabling unreliable transmission it would be inappropriate and unnecessary for this to go through our FIFM. There's no bookkeeping required here once we've sent a QUIC datagram frame.

> Currently, all QUIC implementations which implement RFC 9221 do _some level_ of management, even for unreliable streams, which leads to dropped, or delayed packets. OpenSSL is more flexible in general when it comes to letting users manage connections, so supporting these would go a long way in making this implementation the most optimal for real-time streaming.

A "normal" implementation of RFC 9221 in OpenSSL would probably look like this:

- Datagram payload gets put on a queue
- TX packetiser pulls from queue when generating QUIC packets

No memory copy operations are necessarily added here, it is likely the public API design will give applications the option to avoid the need for that.

But if that turns out to be too indirect for some reason, it does seem conceptually possible to have some kind of direct fastpath in which a SSL_write_datagram call directly corresponds to packet generation and transmission.

> As an aside, would it be possible to extend the BIO API to allow gather for datagrams when writing? Currently, the reference library's entire data path, from a user giving codec packet data to writing it out is entirely zero-copy. Using BIO requires that the datagram contain a single contiguous buffer, which would mean we would need to copy the user data onto a buffer, along with the header the protocol inserts, just so we can pass it off to OpenSSL (which then encodes the data to the output). I am using datagram memory BIO as the reference library implements multiple I/O backends, from standard sendmsg/recvmsg, to using zero-copy liburing (WIP).

Yes, it would be reasonable for an `SSL_write_datagram` API to support gather I/O.

Our intention is for our QUIC stack to support "single-copy" transmission (where the single copy is the act of encrypting data to generate a QUIC packet). This means we will want to provide zero-copy I/O APIs for application data in future (e.g. application provides memory buffer pointer and callback for us to call when we no longer need it).
Of course we could accept a list of iovecs instead of just a single buffer pointer.

This would at least be an SSL_foo API, and may or may not be exposed in the BIO abstraction layer also. The trouble with the latter is that scatter/gather isn't supported on all OSes so we can't support it in e.g. BIO_s_datagram while meeting our portability objectives. So such an API probably wouldn't be supported by a network BIO directly but only a QUIC SSL object (or BIO_f_ssl wrapping it).

**Comment by hlandau** [comment]
> Since it's stateless, sure, that does make sense, as long as no allocations are involved here, which is possible as long as OpenSSL manages the "created" singleton stream objects. Maintaining a constant stream ID is important, so either SSL_write_datagram() would need to accept a stream ID, or a separate function would be needed to set the ID of each singlet. With that, I think a constant SSL stream object makes more sense, but I have no strong feelings about it.

RFC 9221 DATAGRAM frames don't have a stream ID...

Protocols which use the RFC 9221 extension typically put some kind of type field at the beginning of the payload if they need a discriminator.

> Wouldn't this need to be done on a BIO-level?

BIO_f_ssl is just a wrapper around SSL calls, so generally you'd expect there to be a `SSL_write_datagram` function and an exactly matching `BIO_write_datagram` function which BIO_f_ssl would add support for.

It seems likely that it would make sense for `BIO_s_datagram` to also support `BIO_write_datagram` *in principle*. But it probably couldn't support scatter/gather or zero copy portably. My rough expectation is we'd have a copying single-buffer `BIO_write_datagram` API supported by `BIO_s_datagram` and `BIO_f_ssl` and a zero-copy scatter-gather `BIO_write_datagram_zero_copy` or similar (but with a better name) only supported by `BIO_f_ssl` when used with a QUIC connection.

> Glad I can help. There's still a little more work to do to hook everything up together, but I'll keep you updated. So far I haven't ran into issues with the regular reliable stream API, it is well designed and documented. Is there a chat somewhere to discuss in realtime? Couldn't find an official channel in the usual IRC networks.

We don't have an official channel but you can message me (`hl`) on Liberachat or Hackint anytime.

**Comment by mattcaswell** [comment]
> With the release of the server-part of QUIC, are there any plans to move to implementing UDP yet?

It's on our list of things we would like to do, but no concrete plans as yet as to when we will get to it. We would consider external submissions to add such a feature.

**Comment by cyanreg** [comment]
Ping, any plans or progress yet?

**Comment by mattcaswell** [comment]
No progress. Our quic efforts are currently on adding 0-rtt support.
